netty/src/main/java/io/grpc/netty/GrpcHttp2HeadersDecoder.java

```diff
@@ -0,0 +1,498 @@
+    }
+
+    @Override
+    public List<CharSequence> getAll(CharSequence csName) {
+      final AsciiString name = requireAsciiString(csName);
+      final List<CharSequence> returnValues = new ArrayList<CharSequence>(4);
+      for (int i = 0; i < namesAndValuesIdx; i += 2) {
+        if (equals(name, namesAndValues[i])) {
+          returnValues.add(values[i / 2]);
+        }
+      }
+      return returnValues;
+    }
+
+    /**
+     * Returns the header names and values as bytes. An even numbered index contains the
+     * {@code byte[]} representation of a header name (in insertion order), and the subsequent
+     * odd index number contains the corresponding header value.
+     *
+     * <p>The values of binary headers (with a -bin suffix), are already base64 decoded.
+     *
+     * <p>The array may contain several {@code null} values at the end. A {@code null} value an
+     * index means that all higher numbered indices also contain {@code null} values.
+     */
+    byte[][] namesAndValues() {
+      return namesAndValues;
+    }
+
+    /**
+     * Returns the number of none-null headers in {@link #namesAndValues()}.
+     */
+    protected int numHeaders() {
+      return namesAndValuesIdx / 2;
+    }
+
+    protected static boolean equals(AsciiString str0, byte[] str1) {
+      return equals(str0.array(), str0.arrayOffset(), str0.length(), str1, 0, str1.length);
+    }
+
+    protected static boolean equals(AsciiString str0, AsciiString str1) {
+      return equals(str0.array(), str0.arrayOffset(), str0.length(), str1.array(),
+          str1.arrayOffset(), str1.length());
+    }
+
+    protected static boolean equals(byte[] bytes0, int offset0, int length0, byte[] bytes1,
+        int offset1, int length1) {
+      if (length0 != length1) {
+        return false;
+      }
+      return PlatformDependent.equals(bytes0, offset0, bytes1, offset1, length0);
+    }
+
+    protected static byte[] bytes(AsciiString str) {
+      return str.isEntireArrayUsed() ? str.array() : str.toByteArray();
+    }
+
+    protected static AsciiString requireAsciiString(CharSequence cs) {
+      if (!(cs instanceof AsciiString)) {
+        throw new IllegalArgumentException("AsciiString expected.");
+      }
+      return (AsciiString) cs;
+    }
+
+    protected static boolean isPseudoHeader(AsciiString str) {
+      return str.charAt(0) == ':';
+    }
+
+    protected AsciiString validateName(AsciiString str) {
+      final int offset = str.arrayOffset();
+      final int length = str.length();
+      final byte[] data = str.array();
+      for (int i = offset; i < offset + length; i++) {
+        if (isUpperCase(data[i])) {
+          PlatformDependent.throwException(connectionError(PROTOCOL_ERROR,
+              "invalid header name '%s'", str));
+        }
+      }
+      return str;
+    }
+
+    private void expandHeadersAndValues() {
+      int oldLen = namesAndValues.length;
+      byte[][] newHeadersAndValues = new byte[oldLen + oldLen * 2][];
```

Comment:
if the initial length is zero this will fail.

Also ArrayList uses a much less aggressive 1.5 growth.